### PR TITLE
[MRG+1] Added option to disable the webkit in-memory and network caches

### PR DESCRIPTION
**Pull request by wsdookadr**
Hi, I've been reading through the existing open issues.
This pull request is intended to address #203 , #519 , scrapy-plugins/scrapy-splash#168 , #817 , webrecorder/har2warc#1

Please review

Thanks,
Stefan

Note: To try this out, you can use this docker image:

    docker pull wsdookadr/splash:v1
    docker run -it -p 8050:8050  --name splash_v1 wsdookadr/splash:v1 --disable-browser-caches

And then load a web page and generate a .har file from it like so

    curl "http://localhost:8050/render.har?url=https://www.digitalocean.com/community/tutorials/how-to-install-java-with-apt-on-debian-9&timeout=13&wait=4&response_body=1" > /tmp/q.har

If you run this multiple times, the amount of data fetched by curl should be about the same (4mb for that url). Before this pull-request, the first time, it would produce around 4mb, and the subsequent ones about ~200-300k (the .har files would be missing resources that are cached by webkit in its page cache, the first time the page was loaded).

### CODE REVIEW

File: docs/faq.rst [hunk 1]
```diff
@@ -343,6 +343,19 @@ Please refer to `this great answer from kmike on reddit.
...
+
+Webkit has an in-memory cache (also `called page-cache <https://webkit.org/blog/427/webkit-page-cache-i-the-basics/>`)
+and a `network cache <http://doc.qt.io/qt-5/qnetworkrequest.html#CacheLoadControl-enum>`. 
+
+If you tell splash to load two pages that share some common resources,
+the second page's .har file will not contain the shared resources because
+they were cached through the page cache.
+
+If you want the .har file to contain all the resources for that page,
+run splasy with the command-line option ``--disable-browser-caches``.
```
Comment (cathalgarvey) [inline docs/faq.rst]: Minor typo; `splasy` vs. `splash`

File: splash/server.py [hunk 2]
```diff
@@ -345,6 +352,12 @@ def _set_global_render_settings(js_disable_cross_domain_access, private_mode):
     settings.setAttribute(QWebSettings.PrivateBrowsingEnabled, private_mode)
     settings.setAttribute(QWebSettings.LocalStorageEnabled, not private_mode)
 
+    if disable_browser_caches is True:
+        # disables the page cache (also known as the in-memory webkit cache)
+        # see https://webkit.org/blog/427/webkit-page-cache-i-the-basics/
+        settings.setObjectCacheCapacities(0,0,0)
```
Comment (cathalgarvey) [inline splash/server.py]: Minor pep8 gripe; spacing between arguments.

File: docs/faq.rst [hunk 3]
```diff
@@ -343,6 +343,19 @@ Please refer to `this great answer from kmike on reddit.
 
 .. _why-lua:
 
+Why are CSS styling and images missing from the .har archive?
```
Comment (kmike) [inline docs/faq.rst]: the whole section should be moved above `_why-lua:` label, because it is a label for the next section.

Comment (wsdookadr) [reply docs/faq.rst]: yup, moved the section above `_why-lua:` and added its own label

**Comment by kmike** [comment]
Hey @wsdookadr! The PR looks good, but could you please add a test case? Check e.g. [here](https://github.com/scrapinghub/splash/blob/a3f7e757b3bdbc9d3cb6f79c5bfc316a8a05ae69/splash/tests/test_request_filters.py#L149) how a custom startup option is tested. I guess you'd need to define some resource in https://github.com/scrapinghub/splash/blob/a3f7e757b3bdbc9d3cb6f79c5bfc316a8a05ae69/splash/tests/mockserver.py#L975 with caching behavior, then request it 

1) twice with a Splash started with default options - resource should be in har only for one of them,
2) twice with a Splash started with `--disable-browser-caches` - resource should be in har in both cases

**Comment by wsdookadr** [comment]
@kmike I've pushed another commit with tests for this. However, it seems like the problem is not being captured by the test, in other words both default options and the new `--disable-browser-caches` options will bring all the entries with content in the HAR.
The new test can be run using the following command:

    pytest -s --verbosity=6 ./splash/tests/test_request_filters.py -k test_disable_browser_caches

File: splash/tests/mockserver.py [hunk 4]
```diff
@@ -901,6 +901,22 @@ def render_GET(self, request):
             return base64.decodebytes(b'R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=')
 
 
+class SubresourcesWithCaching(Subresources):
+    """ Embedded css and image """
+
+    @use_chunked_encoding
+    def render_GET(self, request):
+        return ("""<html><head>
+                <link rel="stylesheet" href="subresources-with-caching/style.css" />
```
Comment (kmike) [inline splash/tests/mockserver.py]: I think the problem could be that related resources don't set any specific cache headers, so the browser doesn't cache them. Could you try setting [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Caching_static_assets) header for these static resources?

Comment (wsdookadr) [reply splash/tests/mockserver.py]: @kmike you are correct, I've pushed a new commit that sets that Cache-Control header and now it seems that at least for the image.gif resource, the entire scenario is reproducible.

File: splash/tests/test_request_filters.py [hunk 5]
```diff
@@ -155,3 +170,47 @@ def test_file_scheme_can_be_enabled(self):
...
+            entry22 = self.findHAREntry(r.json(),r"^.*subresources-with-caching/style.css$")
+            entry23 = self.findHAREntry(r.json(),r"^.*subresources-with-caching/img.gif$")
+
+            return [[entry11,entry12,entry13],[entry21,entry22,entry23]]
+
+    # run just this test using
+    # pytest -s --verbosity=6 ./splash/tests/test_request_filters.py -k test_disable_browser_caches
+    def test_disable_browser_caches(self):
+        # entries if run with caches enabled
+        # in this case we should have  [[content,content,content],[content,content,no content]]
```
Comment (kmike) [inline splash/tests/test_request_filters.py]: I wonder why isn't it working for css files. Maybe there are some other response headers needed to make it work, or maybe browser uses ETag or something like that, so there is no "full" response. It doesn't look good that opur code sets cache headers, but then checks that the resource is not cached; test will break if resource start to becached.

I think there are two ways to fix that:

1) Investigate what's going on and make sure css is cached as well;
2) remove a check for css file, check only html and gif.

Comment (wsdookadr) [reply splash/tests/test_request_filters.py]: I agree with you on that and I went with the 2nd approach, I have removed the css checks. I've pushed a new commit for this.
For this pull request I'm ok with having it this way.

File: splash/tests/mockserver.py [hunk 6]
```diff
@@ -901,6 +901,51 @@ def render_GET(self, request):
...
+    """
+
+    def getChild(self, name, request):
+        if name == b"style.css":
+            return self.StyleSheet()
+        if name == b"img.gif":
+            return self.Image()
+        return self
+
+    class StyleSheet(Resource):
```
Comment (kmike) [inline splash/tests/mockserver.py]: Can it be removed now, as well as `<link rel=stylesheet>` tag in html?

File: splash/tests/mockserver.py [hunk 7]
```diff
@@ -901,6 +901,39 @@ def render_GET(self, request):
             return base64.decodebytes(b'R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=')
 
 
+class SubresourcesWithCaching(Resource):
+    """
+        Embedded css and image.
```
Comment (kmike) [inline splash/tests/mockserver.py]: docstring is no longer correct, as it embeds only image

File: splash/tests/mockserver.py [hunk 8]
```diff
@@ -901,6 +901,39 @@ def render_GET(self, request):
...
+        @use_chunked_encoding
+        def render_GET(self, request):
+            request.setHeader(b"Content-Type", b"image/gif")
+            request.setHeader(b"Cache-Control", b"public, max-age=999999, s-maxage=999999")
+            return base64.decodebytes(b'R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=')
+
+    @use_chunked_encoding
+    def render_GET(self, request):
+        return ("""<html><head>
+            </head>
```
Comment (kmike) [inline splash/tests/mockserver.py]: I think you can remove `<head>`, to make an example a bit shorter.

**Comment by kmike** [comment]
Thanks @wsdookadr! 🎉

**Comment by Mideen** [comment]
Hi team,
 When can we expect the next stable version with these changes? 
@kmike

**Pull request merged**